prov/mrail/src/mrail_cq.c

```diff
@@ -51,6 +82,143 @@ int mrail_cq_write_recv_comp(struct mrail_ep *mrail_ep, struct mrail_hdr *hdr,
+{
+	FI_DBG(&mrail_prov, FI_LOG_CQ, "finish rndv recv: length: %zu "
+	       "tag: 0x%" PRIx64 "\n", recv->rndv.len, recv->rndv.tag);
+	ofi_ep_rx_cntr_inc(&mrail_ep->util_ep);
+	if (!(recv->flags & FI_COMPLETION))
+		return 0;
+	return ofi_cq_write(mrail_ep->util_ep.rx_cq, recv->context,
+			   recv->comp_flags | recv->rndv.flags,
+			   recv->rndv.len, NULL, recv->rndv.data,
+			   recv->rndv.tag);
+}
+
+static void mrail_finish_rndv_recv(struct mrail_req *req,
+				   struct fi_cq_tagged_entry *comp)
+{
+	int ret;
+	struct mrail_recv *recv = req->comp.op_context;
+
+	ret = mrail_cq_write_rndv_recv_comp(req->mrail_ep, recv);
+	if (ret) {
+		FI_WARN(&mrail_prov, FI_LOG_CQ,
+			"Cannot write to recv cq\n");
+		assert(0);
+	}
+
+	do {
+		ret = mrail_send_rndv_ack(req->mrail_ep, recv->addr,
+					  (void *)recv->rndv.context);
+		if (ret && ret != -FI_EAGAIN) {
+			FI_WARN(&mrail_prov, FI_LOG_CQ,
+				"Cannot send rndv ack: %s\n", fi_strerror(-ret));
+			assert(0);
+		}
+	} while (ret == -FI_EAGAIN);
+
+	mrail_free_req(req->mrail_ep, req);
+	mrail_push_recv(recv);
+}
+
+static int mrail_cq_process_rndv_req(struct fi_cq_tagged_entry *comp,
+				     struct mrail_recv *recv)
+{
+	struct fi_recv_context *recv_ctx = comp->op_context;
+	struct fi_msg msg = {
+		.context = recv_ctx,
+	};
+	struct mrail_ep *mrail_ep;
+	struct mrail_pkt *mrail_pkt;
+	struct mrail_rndv_hdr *rndv_hdr;
+	struct mrail_rndv_req *rndv_req;
+	struct fi_msg_rma rma_msg;
+	uint64_t *base_addrs;
+	size_t key_size;
+	size_t offset;
+	int ret, retv = 0;
+	int i;
+
+	mrail_ep = recv_ctx->ep->fid.context;
+	mrail_pkt = (struct mrail_pkt *)comp->buf;
+	rndv_hdr = (struct mrail_rndv_hdr *)&mrail_pkt[1];
+	rndv_req = (struct mrail_rndv_req *)&rndv_hdr[1];
+	recv->rndv.context = (void *)rndv_hdr->context;
+	recv->rndv.flags = comp->flags & FI_REMOTE_CQ_DATA;
+	recv->rndv.len = rndv_req->len;
+	recv->rndv.tag = mrail_pkt->hdr.tag;
+	recv->rndv.data = comp->data;
+
+	base_addrs = (uint64_t *)(rndv_req->rawkey + rndv_req->rawkey_size);
+	for (offset = 0, i = 0; i < rndv_req->count; i++) {
+		if (i < rndv_req->mr_count) {
+			key_size = rndv_req->rma_iov[i].key;
+			ret = fi_mr_map_raw(&mrail_ep->util_ep.domain->domain_fid,
+					    base_addrs[i],
+					    rndv_req->rawkey + offset,
+					    key_size,
+					    &rndv_req->rma_iov[i].key,
+					    0);
+			assert(!ret);
+			offset += key_size;
+		} else {
+			rndv_req->rma_iov[i].key = rndv_req->rma_iov[0].key;
+		}
+	}
+
+	rma_msg.msg_iov		= recv->iov + 1;
+	rma_msg.desc		= recv->desc + 1;
+	rma_msg.iov_count	= recv->count - 1;
+	rma_msg.addr		= recv->addr;
+	rma_msg.rma_iov		= rndv_req->rma_iov;
+	rma_msg.rma_iov_count	= rndv_req->count;
+	rma_msg.context		= recv;
+
+	ret = fi_readmsg(&mrail_ep->util_ep.ep_fid, &rma_msg,
+			 MRAIL_RNDV_FLAG | FI_COMPLETION);
+	if (ret)
+		retv = ret;
+
+	ret = fi_recvmsg(recv_ctx->ep, &msg, FI_DISCARD);
+	if (ret) {
+		FI_WARN(&mrail_prov, FI_LOG_CQ,
+			"Unable to discard buffered recv\n");
+		retv = ret;
```

Comment:
use
```
retv = retv ? retv : ret;
```
?

Reply:
That's probably more desirable since the first error code is more important.